Document logging architecture (#2170)
giantswarm/docs: src/content/vintage/getting-started/observability/logging/architecture/index.md

---
linkTitle: Logging architecture
title: Logging architecture
description: Documentation on the logging architecture deployed and maintained by Giant Swarm.
weight: 80
menu:
  main:
    identifier: getting-started-observability-logging-architecture
    parent: getting-started-observability-logging
user_questions:
  - What is the logging architecture?
  - Why is Giant Swarm using loki?
  - Why is Giant Swarm recommending loki?
  - Which logs are stored by Giant Swarm?
  - Where are the logs stored by Giant Swarm?
aliases:
  - /getting-started/observability/logging/architecture
owner:
  - https://github.com/orgs/giantswarm/teams/team-atlas
last_review_date: 2024-03-21
---

> Logging is an important pillar of observability and so it is expected that Giant Swarm provides and manages a logging solution for operational purposes.

Reviewer: ```suggestion
Logging is an important pillar of observability and it is thus only natural that Giant Swarm provides and manages a logging solution for operational purposes.
```

> This documentation will teach you about which logs are stored by Giant Swarm, which technology we use to ship and store them as well as why we use those tools.

Reviewer: ```suggestion
This documentation will give you an overview of how logging is managed by Giant Swarm : which logs are stored, which tools we use to ship and store them as well as why we chose those tools in the first place.
```

## Overview of the logging pipeline

Logging agents, Loki and grafana, log pipeline overall.

## Logs stored by Giant Swarm

Kubernetes clusters produces a vast amount of logs, whether they come from machines or containers.

The loggings agents that we have deployed on both management and workload clusters currently send the following logs to Loki:
- `Kubernetes container logs` in the `kube-system` and `giantswarm` namespaces.
- `Kubernetes events` happenning in the `kube-system` and `giantswarm` namespaces.
- [`Kubernetes audit logs`]({{< relref "./audit-logs#kubernetes-audit-logs" >}})

In the future, we will also store the following logs:
- [`Machine audit logs`]({{< relref "./audit-logs#machine-audit-logs" >}})
- `Teleport audit logs` currently being worked on here: https://github.com/giantswarm/roadmap/issues/3250
- Giant Swarm `customer workload logs` as part of our observability platform being worked on https://github.com/giantswarm/roadmap/issues/2771

## Why we prefer Loki over its competitors

> The reasons that lead us to using Grafana Loki in favor of its competitors (which is mostly opendistro in our case) are numerous.

Reviewer: ```suggestion
The reasons that lead us to using Grafana Loki in favor of its competitors (which boil down to only opendistro in our case) are numerous.
```

First, we are **strong believers in Open Source** so the full Elastic stack is obviously out of the question.

> Second, we are quite used to the Grafana ecosystem and the different tools are made to work with one another whereas the existing logging solutions are either supposed to work on their own (like OpenDistro) or you need to use the full fledge solution (i.e. being able to **collect and correlate all observability data**) which is rarely open-source (coming back to the first point above).

Reviewer: ```suggestion
Second, we are quite used to the Grafana ecosystem and the different tools are made to work with one another whereas the existing logging solutions are either supposed to work on their own (like OpenDistro) or need to use the full fledged solution (i.e. being able to **collect and correlate all observability data**) which is rarely open-source (coming back to the first point above).
```

> Third, we are full-fledge users of Prometheus and PromQL and **LogQL, the Loki Query Language is a natural extension to PromQL**, which makes it easy for our platform engineers to use and love.

Reviewer: ```suggestion
Third, we are full-fledged users of Prometheus and PromQL and **LogQL, the Loki Query Language is a natural extension to PromQL**, which makes it easy for our platform engineers to use and love.
```

> The fourth reason is **cost and resource consumption** as Loki is cheaper to run than because it does not rely that much on persistent storage and uses Object storage instead which is always cheaper in the cloud. The storage of the index is also cheaper for Loki as it uses label-based indexing which is definitely smaller than any kind of text-based indexing solution that full-text search engine can provide.

Reviewer: ```suggestion
The fourth reason is **cost and resource consumption** as Loki is cheaper to run than its competitors because it does not rely that much on persistent storage and uses Object storage instead which is always cheaper in the cloud. The storage of the index is also cheaper for Loki as it uses label-based indexing which is definitely smaller than any kind of text-based indexing solution that full-text search engine can provide.
```

> Finally, the last reason comes down to the history of Giant Swarm and it mostly boil down to **operation and maintenance**. Before we decided to run Loki, we used to run elasticsearch as our logging solution. Elasticsearch in itself is really hard to operate, especially at scale, even more so on Kubernetes because it is by its nature a stateful application (and for good reasons). This was an especially important factor in our decision since we do not need the full capabilities of OpenDistro like full-text search.

Reviewer: ```suggestion
Finally, the last reason comes down to the history of Giant Swarm and it mostly boils down to **operation and maintenance**. Before we decided to run Loki, we used to run elasticsearch as our logging solution. Elasticsearch in itself is really hard to operate, especially at scale, even more so on Kubernetes because it is by its nature a stateful application (and for good reasons). This was an especially important factor in our decision since we do not need the full capabilities of OpenDistro like full-text search.
```

---
linkTitle: Logging architecture
title: Logging architecture
description: Documentation on the logging architecture deployed and maintained by Giant Swarm.
weight: 80
menu:
  main:
    identifier: getting-started-observability-logging-architecture
    parent: getting-started-observability-logging
user_questions:
  - What is the logging architecture?
  - Why is Giant Swarm using Loki?
  - Why is Giant Swarm recommending Loki?
  - Which logs are stored by Giant Swarm?
  - Where are the logs stored by Giant Swarm?
aliases:
  - /getting-started/observability/logging/architecture
owner:
  - https://github.com/orgs/giantswarm/teams/team-atlas
last_review_date: 2024-03-21
---

Logging is an important pillar of observability and it is thus only natural that Giant Swarm provides and manages a logging solution for operational purposes.

> This document gives an overview of how logging is managed by Giant Swarm, including which logs are stored, which tools we use to ship and store them, as well as why we chose those tools in the first place.

Reviewer 2: This doesn't tell the customer much why it's helpful for them. What are operational purposes? Does that mean they or only we can make use of it? I'd recommend shortening this intro, include customer-relevant points and only describe what's important to them. I assume that customers can make use of the MC Grafana instance (see existing article `How to access Grafana` which btw has a misspelled title `it's`/`its`) and therefore we should strongly hint here why the architecture is important for customers (e.g. long-term storage, persistence, easy access via Grafana UI, ...) but not go into extensive detail or opinions ("the full Elastic stack is obviously out of the question").

@pipo02mix I'm very unclear what was discussed around vintage docs. Shouldn't we start from scratch rather than editing existing docs? If so, we'd better revamp the whole thing and go through the "battle card" preparation before writing full articles. Since I'm unclear about this, I'd like to have it clarified first and will stop my review here for now.

Author: I may be biased but operational purposes is quite straightforward to me. We need it to operate our platform and those 2 sentences have the same meanings. 
I am not talking about the customer use yet here, only that they can access it via grafana.

Now, maybe the why we use Loki instead of Grafana could be somewhere else, I wrote this under the after some discussions with @carillan81 and @pipo02mix .

The goal of this doc is to show what we have and show our expertise but maybe the battlecard approach was missed.

Regarding the structure, I put this under vintage for now because of the full revamp being done and I have no issues moving it and refining the content later on with said battlecard, but in the mean time, I think this type of doc is useful and needs to be there. I could be wrong also but it's still better than having nothing at the moment at least

Reviewer 2: Agreed – I was writing my direct thoughts to challenge and make everyone think if everything still fits what we have here. We should check from scratch which pieces we put where, and if we can make it all shorter to increase chances of reading these articles. For example, does it fit into introductional, instructional or rather advanced parts of our docs, etc.?

Author: I'm always happy with any feedback :)

If you want to play with Loki, you should definitely check out our guides explaining [how to access Grafana]({{< relref "/vintage/getting-started/observability/visualization/access" >}}) and how to [explore logs with LogQL]({{< relref "/vintage/getting-started/observability/visualization/log-exploration" >}})

## Logs stored by Giant Swarm

Kubernetes clusters produce a vast amount of machine and container logs.

The logging agents that we have deployed on management and workload clusters currently send the following logs to Loki:

- Kubernetes Pod logs from the `kube-system` and `giantswarm` namespaces.
- Kubernetes Events created in the `kube-system` and `giantswarm` namespaces.
- [Kubernetes audit logs]({{< relref "./audit-logs#kubernetes-audit-logs" >}})

In the future, we will also store the following logs:

- [Machine (Node) audit logs]({{< relref "./audit-logs#machine-audit-logs" >}})
- Teleport audit logs, tracked in https://github.com/giantswarm/roadmap/issues/3250
- Giant Swarm customer workload logs as part of our observability platform, tracked in https://github.com/giantswarm/roadmap/issues/2771

## Why we prefer Loki over its competitors

> The reasons that lead us to using Grafana Loki in favor of its competitors (which boils down to only opendistro in our case) are numerous.

Reviewer 3: ```suggestion
There are numerous reasons to use Grafana Loki in favor of its competitors.
```
Just less cooking here, because we already use that expression and also mention OpenDistro below.

First, we are **strong believers in Open Source** so the full Elastic stack is obviously out of the question.

> Second, we are quite used to the Grafana ecosystem and the different tools are made to work with one another whereas the existing logging solutions are either supposed to work on their own (like OpenDistro) or need to use the full-fledged solution (i.e. being able to **collect and correlate all observability data**) which is rarely open-source (coming back to the first point above).

Reviewer 4: ```suggestion
Second, we are quite used to the Grafana ecosystem, where the **individual tools are made to work with one another without requiring a closed ecosystem**. Alternative logging solutions are either intended to work in isolation (like OpenDistro) or need to use a full-fledged solution (i.e. being able to collect and correlate all observability data), which is rarely open-source (coming back to the first point above).
```
Changed emphasis here to better match the other paragraphs

> Third, we are full-fledged users of Prometheus and PromQL and **LogQL, the Loki Query Language is a natural extension to PromQL**, which makes it easy for our platform engineers to use and love.

Reviewer 4: ```suggestion
Third, we are full-fledged users of Prometheus and PromQL. **LogQL, the Loki Query Language, is a natural extension to PromQL**, which makes it easy for our platform engineers to use and love.
```

> The fourth reason is **cost and resource consumption** as Loki is cheaper to run than its competitors because it does not rely that much on persistent storage and uses Object storage instead which is always cheaper in the cloud. The storage of the index is also cheaper for Loki as it uses label-based indexing which is definitely smaller than any kind of text-based indexing solution that full-text search engine can provide.

Reviewer 4: ```suggestion
The fourth reason is **cost and resource consumption.** Loki is cheaper to run than its competitors because it does not rely as heavily on persistent storage and uses Object storage instead, which is always cheaper in the cloud. The storage of the index is also cheaper for Loki as it uses label-based indexing, which is smaller than any kind of text-based indexing solution used by full-text search engines.
```
